### PR TITLE
API enhancements

### LINKED ISSUE #103: API enhancements
- [x] refresh file for viewers from disk with POST request (needed for Jupyter-from-Vim-project)
- [x] some way to open a viewer at a given scroll position (needed for [vivify.vim#6](https://github.com/jannis-baum/vivify.vim/issues/6)
- [x] remove ability to configure port through config file (only env from now on)

### PR DESCRIPTION
**Pull request by jannis-baum**
Closes #103

### CODE REVIEW

**Comment by jannis-baum** [comment]
The scrolling is getting more complicated than I thought🥲 @tuurep can you help me with something?

1. What happens when you turn off the redirection to `~` and run `viv some_file` and then run it again on the current version (not this PR)? Does it open 2 tabs or does it open one and then the second time just focus the one that was already there?
2. What happens when you use `xdg-open` with the correct URL and do that twice? Again, 2 tabs or 1 tab+focus that tab?

**Comment by tuurep** [comment]
Oh no :D

Noted, will be getting back to this another day.

**Comment by jannis-baum** [comment]
I don't need your answers to all those questions anymore, thanks anyways! I found a solution, I think this is ready for testing & review now :)

**Comment by tuurep** [comment]
`viv --scroll` / `-s` seems like a fantastic idea

So is the meaning of `n`, that the content of `n` is visible at the bottom of page? Like for example I do `viv -s 10 README.md` on the viv repo README, it doesn't scroll because line 10 is already visible from the top?

If I understand that correctly, then it seems to work

> It should not open a second tab if you already have a viewer open, instead it should focus that tab and scroll to the position on it.

This however doesn't quite work, it does the scroll seemingly correctly on the existing tab, but then also opens a new tab with scroll at the top

**Comment by jannis-baum** [comment]
> So is the meaning of n, that the content of n is visible at the bottom of page? Like for example I do viv -s 10 README.md on the viv repo README, it doesn't scroll because line 10 is already visible from the top?

It's the same thing that the POST request from Vim does: It scrolls the minimal possible amount so that the content corresponding to line `n` is in the viewport. I.e. when you newly open the page it will most likely end up at the bottom of the page.

> This however doesn't quite work, it does the scroll seemingly correctly on the existing tab, but then also opens a new tab with scroll at the top

Shit ^^ what happens when you `xdg-open` a URL that is already open in some tab? Does it also open a new tab anyways? (not in particular for Vivify-URLs but just in general, also with e.g. GitHub or something)

> When I keep closing tab and doing viv -s 94 README.md repeatedly, it appears if it's working every other time (half the time not scrolling at all)

Seems like you were very fast there! Up until now the clients only got cleaned up after not replying to a `ping`, i.e. if you closed the tab the server needed up to 2 seconds to know and be able to properly handle the scrolling. I just added something that also clears them when the socket is properly closed so it should always work almost immediately now :)

**Comment by jannis-baum** [comment]
Ah, and adding to the `xdg-open` topic: Is there some way you could make your browser open the existing tab from the command line based on only the URL? E.g. `mybrowser --fancy-option-that-reuses-tabs http://some/url/that/is/already/open/in/a/tab`

**Comment by tuurep** [comment]
Yeah it opens a new tab with `xdg-open` (or `firefox`)

So must be a firefox specific thing

As far as I can see with a quick search, it looks like there's no flag for that

Maybe achievable with some scripting but hard to say

**Comment by jannis-baum** [comment]
> As far as I can see with a quick search, it looks like there's no flag for that

Hm, that's sad. I feel like it would be quite a significant UX-flaw not to reuse already open tabs. On macOS and Safari it works.

This also puts us into a dilemma of how to implement this because we can only *either* have the reusing work for systems where you can open already open tabs, *or* open a new one at the given scrolling position. Or of course we can also support both with an option, but then it'll be a super-hard-to-explain option that is easy to set incorrectly or not set at all.

One other approach I see is to not use `open` at all but try focussing the tab from the client-side JS. We can try and investigate if that's blocked by our browsers or if it works, but I imagine Safari for example will block it to be honest.

**Comment by jannis-baum** [comment]
Just for context, I found this

- Firefox setting that may be relevant: https://support.mozilla.org/en-US/questions/1226151
- Chrome extension that might do it: https://chromewebstore.google.com/detail/one-tab-only/cdkbmeoilbjdnbaonjnenjdafkgaeabo?hl=en

Other than that it would probably be possible with some scripting. I don't know how deep we want to get into this, especially since every combination of system/browser/window-manager might need its own script. What do you think about this @tuurep and what browser and window manager do you use?

**Comment by tuurep** [comment]
The firefox thing looks like it's not related to the URL being the same -case and I don't think it's usable for our case

Chrome extension looks very relevant however

Yeah it seems really difficult, even more so with the X/Wayland divide

I'm using Openbox window manager and Firefox

**Comment by jannis-baum** [comment]
Just saw this as well: https://github.com/balta2ar/brotab. That might actually cover a majority of use cases.

Anyways, if anything I would revert Vivify to give the option to specify an `open`-command so that people could (for example) put a script based on that tool there and have it reuse tabs. Is this something you would do if the option was there? Or is this topic not that important to you anyways?

**Comment by jannis-baum** [comment]
> > https://github.com/balta2ar/brotab
> 
> Huh, might be the coolest thing I've seen in a while

Since Vivify I assume :p

> Well, I'm not used to tab reuse in any other context than this but it interests me
> 
> By far I'm most interested in the scrolling to the correct location on invoking `:Vivify` from nvim, but are these two cases related?

Yes, 100% related. The way we would do it from Vim is by running `viv --scroll`.

**Comment by tuurep** [comment]
> Since Vivify I assume :p

Vivify is TheBest

> The way we would do it from Vim is by running `viv --scroll`.

Yeah but does this mean that all cursor-page syncing is migrated to doing `viv --scroll`s?

Because for that case it's enough for me that `:Vivify` opens a new tab, no need to reuse a previous one (and not relevant)

**Comment by jannis-baum** [comment]
> > The way we would do it from Vim is by running `viv --scroll`.
> 
> Yeah but does this mean that all cursor-page syncing is migrated to doing `viv --scroll`s?
 
No. The POST request is for scrolling whatever is already open, `viv --scroll` is for opening and then scrolling.

> Because for that case it's enough for me that `:Vivify` opens a new tab, no need to reuse a previous one (and not relevant)

Okay! If you think this is not that necessary then we can just not have it and the only problem we have is having to guess if our `open` command will be able to reuse tabs or not.

**Comment by tuurep** [comment]
Although yeah there was this issue:

When I open same file twice with `-s 94`, the first scrolls correctly, but the second scrolls the first tab, and then opens a new tab at the top

When my expectation would be: don't (re)scroll the existing one, only open the new tab and then scroll on that

**Comment by jannis-baum** [comment]
> When I open same file twice with -s 94, the first scrolls correctly, but the second scrolls the first tab, and then opens a new tab at the top
>
> When my expectation would be: don't (re)scroll the existing one, only open the new tab and then scroll on that

Yes, exactly, this is the issue with your browser not reusing the tab. If it would, then it would not open the second tab and instead focus the first one, which also scrolled. That's how it's intended

**Comment by jannis-baum** [comment]
I think I actually just figured out a different solution:

Instead of opening the URL to the requested file from the second `vivify-server` instance that will terminate and just serves as a CLI, we can send a POST request from there to the first (running server) instance and then have it open the URL itself and at the same time register something that will send the `SCROLL` message to the first (next) socket that connects at that path. Then it works no matter if the tab is reused or not and we don't have to do any ugly hacks.

**Comment by tuurep** [comment]
Alright! I'll be back on this one, but I wanna go and do https://github.com/jannis-baum/vivify.vim/issues/5 right now

**Comment by jannis-baum** [comment]
Hey @tuurep this should finally be good now!

Aside from removing configurability of the port through the config file as we discussed on the Vim plugin, we introduce two new features:

1. We can now post a `reload` to the viewer route with should make the viewer live (without an actual browser-reload) update by reading the corresponding file again.
2. The `--scroll`/`-s`. This part is more complicated, I will explain how I did it now below.

The URL query parameter approach isn't feasible because it wouldn't allow us to make us of browsers who can reuse tabs like Safari (opening a URL with query parameters is not the same as the already open tab that doesn't have any parameters).

Instead, we introduce message queuing in the `src/sockets.ts` file. Here, we keep a mapping from paths to messages that will be sent the next client that connects at the given path. This way, we can queue the `SCROLL` message to be sent right after the viewer opens, which happens both for browsers that do and that don't reuse tabs (the tab reloads anyways).

One final caveat is that we can't simply queue the message in the process we're currently in, because a new Vivify process will reuse an already running server and exit itself. To handle this I, introduced a new route at `/_open` which should only be used internally. It's basically just an HTTP interface to queueing a message and then opening a viewer so that this can happen on whatever process (itself or an older one) is handling the server.[^1]

I'd be happy if you could test this and potentially also look through the code if it's not too much trouble! Thank you!

[^1]: Of course this approach comes with the limitation that it would break if we run `viv` at the exact same time twice with at least one of them having a scroll target. I would argue that this will never happen in practice, unless someone exposes the Vivify server to some network and actually has many clients. I would argue that this would currently be very un-smart anyways, since it would expose the entire file system, but if we ever wanted to support this we would have to come up with some smart synchronization between client and server that protects the critical section between queuing the message and the next client connecting. I added a comment in the code to mention this so we don't forget. But like I said, I don't think this is relevant right now.

**Comment by tuurep** [comment]
Yes now this feels like it works!

There's some very complicated stuff and I'm glad you've written it out, but I don't think I should try to fully understand it all at this point, but might do if problems arise out of this.

I took a general look at all the code, looks good.

I only have one idea about the CLI:

It doesn't allow working with filenames starting with `-`. Which is quite obvious lol, but: there's a common way tools often work around this problem, which is a double dash:

https://unix.stackexchange.com/questions/11376/what-does-double-dash-double-hyphen-mean

For example if I wanna create and edit a file named `-footgun.md`, I can do:

`$ nvim -- -footgun.md`

So just bringing this up as reference if any kinda dash-prefixed non-flag arguments become a problem. This is such an edge case at the moment I don't think we need to do anything about it at this point :smile: Unless, for example, that would be super easy to add and you wanna.

Anything else to test? Seems fine to me. I took a preliminary look at https://github.com/jannis-baum/vivify.vim/pull/17 and looks great how the plugin needs only a tiny change thanks to this. Going to try that.

**Comment by jannis-baum** [comment]
Awesome, thank you so much for looking at this!

Great idea about the `--`! I added it, it wasn't a bit adjustment :) If you can confirm that this works I think we are ready to merge🥳

**Comment by tuurep** [comment]
Oh hold on

Looks like multiple file arguments i.e. `viv a.md b.md` isn't acting quite right

It's only able to open one, and for the others gives a 

```
Duplicate argument for "target", skipping
```

This is unintended, right?

However wow the `--` works perfect, thought it would involve quite a bit more :)

**Comment by jannis-baum** [comment]
> Oh hold on
> 
> Looks like multiple file arguments i.e. `viv a.md b.md` isn't acting quite right
> 
> It's only able to open one, and for the others gives a
> 
> ```
> Duplicate argument for "target", skipping
> ```
> 
> This is unintended, right?

Nope, this is intended😂😂 I couldn't come up with a good way of figuring out which target the `--scroll` argument(s) correspond to without making things ugly and complicated so I got rid of the feature of opening multiple files lol. Do you think that's an issue? Definitely seems like the error message is not good ^^

**Comment by tuurep** [comment]
Oohh.. okay

It's a bit of a drawback, would be great if opening in bulk worked, but there's need for a compromise I see.

Yeah the error message wasn't understandable to me at least

**Comment by jannis-baum** [comment]
> It's a bit of a drawback, would be great if opening in bulk worked, but there's need for a compromise I see.

It's more of a question of how the interface should be. How would you expect this to work? E.g. if I run

```txt
viv a.md --scroll 50 b.md
```

What happens? Which file scrolls?

My initial implementation was to have something like

```txt
viv a.md b.md:50
```

which would be beautiful in a dream world where all the evil characters are forbidden in paths😂 but of course that's not the reality haha

**Comment by tuurep** [comment]
> viv a.md b.md:50

Ah, yes this would be so elegant :smile: Is it not possible to just require that `:`'s are escaped like `\:`?

> viv a.md --scroll 50 b.md

For this yeah... Well I think I'd expect all files to scroll and that I'm using it in a weird way on purpose

**Comment by tuurep** [comment]
Btw interesting and related, from `man nvim`:

```
+[linenum]  For  the  first  file, position the cursor on line linenum.  If linenum is
            omitted, position the cursor on the last line of the file.  +5 and -c 5 on
            the command-line are equivalent to :5 inside nvim.
```

Example:

`$ nvim +100 README.md LICENSE`

Only scrolls on README

**Comment by jannis-baum** [comment]
> Ah, yes this would be so elegant 😄 Is it not possible to just require that `:`'s are escaped like `\:`?

Oh, you might be on to something here! I like this, wanna go for this one?

> > viv a.md --scroll 50 b.md
> 
> For this yeah... Well I think I'd expect all files to scroll and that I'm using it in a weird way on purpose

Yeah, I also considered this but I don't think it's a useful feature to scroll all files to the same source line.

> Btw interesting and related, from man nvim:

Hm, yes, we could also do this but I think the escaping the colon is nicer. What do you think?

**Comment by tuurep** [comment]
Yeah

> I don't think it's a useful feature to scroll all files to the same source line.

Agreed definitely

> escaping the colon

Yes if this is possible, would be really cool!

**Comment by jannis-baum** [comment]
> Yes if this is possible, would be really cool!

Cool! I'll have a look into that soon then. Should be rather straight forward I think haha but I am wrong with this statement often enough so let's see :D

**Comment by jannis-baum** [comment]
Hello @tuurep so I implemented this with a monster of a regular expression :D 

https://github.com/jannis-baum/Vivify/blob/13827444c123e9460820611f7725877ff0f144cd/src/cli.ts#L11

After debating quite a bit, this seemed like the best way to do it because it is very robust, fast and can cover lots of scenarios without writing tons of complicated code. With this expression we require an absolute minimal amount of escaping.

However, since there is absolutely no way to expect anyone to understand or manually test that expression, I decided to make unit tests for it in `tests/unit/cli.ts`. Would be great if you could have a look at those. Feel free to thin them out, adjust them, and/or add more :)

This PR is ready for re-review now!

File: src/parser/config.ts [hunk 1]
```diff
@@ -2,6 +2,8 @@ import fs from 'fs';
 import { homedir } from 'os';
 import path from 'path';
 
+// NOTE: this does type not directly correspond to the config file: see
```
[the remark this answers is not in the document]
Comment (jannis-baum) [reply src/parser/config.ts:5]: yes lol, will fix that later!

**Comment by tuurep** [comment]
Hey, yep this works and overall fantastic

The only thing I can notice but I'm not sure of the expected behavior:

If you bulk-open the same file many times with different lines to scroll i.e.

`$ viv README.md:50 README.md:100`

only one scroll, other opens at the top. What do you think about that?

When files are different, both scroll. Also tested with 3 different and works.

The tests are very comfortable and I got them to run and pass first try by only glancing at what you documented about them.

Not sure what tests to add but noted for the future.

**Comment by jannis-baum** [comment]
> If you bulk-open the same file many times with different lines to scroll i.e.
> 
> $ viv README.md:50 README.md:100

Yes, this is directly caused by what I wrote in the footnote [here](https://github.com/jannis-baum/Vivify/pull/117#issuecomment-2254005545) and in the comment

https://github.com/jannis-baum/Vivify/blob/6237788221b6ceb7010d48b1ebad6be1d6e102a9/src/routes/_open.ts#L18-L21

We could "fix" this by adding something like a mutex or semaphore around the section I described in the comment but I am not sure how much sense this makes since live updates from an editor for example will always affect all clients (tabs) the same way, i.e. all will end up at the exact same state.

I have never done low level synchronization like this in Node/JavaScript so I'm not sure how much complexity it would introduce but I can definitely look into it. What do you think?

**Comment by tuurep** [comment]
Hmmm.. Do I read into it correctly that it could be a bit of a nightmare for a small detail? :smile:

The use case of opening the same file at 2 places is sensible, but maybe not quite important. In fact could be one of our edgiest edge cases.

**Comment by jannis-baum** [comment]
> Hmmm.. Do I read into it correctly that it could be a bit of a nightmare for a small detail? 😄

Maybe haha. I can also just set myself a 10 minute timer and see how far I get and then decide :D Maybe I'll do that, we have to see how thread-safe objects in JS are and how/if Express/NodeJS even implements multithreading. If it's all just one thread it will be very easy.

**Comment by tuurep** [comment]
Ok sounds good!

**Comment by jannis-baum** [comment]
So… I have said this before but I think/hope this should finally be all done now ^^ The message queuing is a lot more robust this way I think, I just can't really test opening the same file multiple times at different scroll locations because of course Safari reuses the same tab for all requests ^^ So let me know if this works now :)

**PS**: If you want to look at what I changed to make this (hopefully) work now, you can just look at `808cd17`, e.g. with `git diff 808cd17^ 808cd17`; after that I just refactored some stuff to make things easier to understand.

**Comment by tuurep** [comment]
Great!

Yes it works, for example 

```
$ viv README.md:60 README.md:100 README.md:70
```

all as expected.

> `git diff 808cd17^ 808cd17`

Thanks, I looked at it but hopefully won't have to go deep into it :smile: But sounds like good changes generally as well.

So, do you think it can be merged?

**Comment by jannis-baum** [comment]
Great, thanks! Yes, let's finally merge this
